reinforced_lib/agents/particle_filter.py

```diff
@@ -0,0 +1,283 @@
+from functools import partial
+from typing import Tuple
+
+import chex
+import distrax
+import gym.spaces
+import jax
+import jax.numpy as jnp
+from jax.scipy.special import erf
+
+from reinforced_lib.agents.agent_state import AgentState
+from reinforced_lib.agents.base_agent import BaseAgent
+
+
+@chex.dataclass
+class ParticleFilterState(AgentState):
+    """
+    Container for the state of the Particle Filter agent.
+
+    Fields
+    ------
+    positions : chex.Array
+        Positions of filter particles.
+    weights : chex.Array
+        Weights of filter particles.
+    last_sample : float
+        Time of the last sampling.
+    """
+
+    positions: chex.Array
+    logit_weights: chex.Array
+    last_sample: chex.Scalar
+
+
+def init(
+        min_position: chex.Scalar,
+        max_position: chex.Scalar,
+        particles_num: jnp.int32
+) -> ParticleFilterState:
+    """
+    Creates and initializes instance of the Particle Filter agent.
+
+    Parameters
+    ----------
+    min_position : float
+        Minimum position of a particle.
+    max_position : float
+        Maximum position of a particle.
+    particles_num : int
+        Number of created particles.
+
+    Returns
+    -------
+    out : ParticleFilterState
+        Initial state of the Particle Filter agent.
+    """
+
+    return ParticleFilterState(
+        positions=jnp.linspace(min_position, max_position, particles_num),
+        logit_weights=jnp.zeros(particles_num),
+        last_sample=0.0
+    )
+
+
+def update(
+        state: ParticleFilterState,
+        key: chex.PRNGKey,
+        action: jnp.int32,
+        n_successful: jnp.int32,
+        n_failed: jnp.int32,
+        time: chex.Scalar,
+        power: chex.Scalar,
+        cw: jnp.int32,
+        min_snr: chex.Array,
+        scale: chex.Scalar
+) -> ParticleFilterState:
+    """
+    Updates the state of the agent after performing some action and receiving a reward.
```

Comment:
Zacząłem się zastanawiać czy nie warto jakoś bardziej szczegółowo te metody agentów opisywać. Chodzi mi o te opisy przed parametrami. Bo wszędzie (narazie w PF i TS) mamy takie generyczne opisu typu 'wybiera kolejną akcję i zwraca nagrodę', a może warto by tutaj uściślić czym ta akcja jest (np że jest to kolejny MCS) i co konkretnie jest tą nagrodą (np sukces/porażka). No i ten komentarz odnosi się do obu tych funkcji `update()` oraz `sample()`

Reply:
Hmm, w sumie to jest dość dobra uwaga. Na razie pozostawiałem takie generyczne opisy, bo nie miałem zbyt wiele czasu na dorobienie dokładnego opisu, ale ogólnie, że to bardzo dobry pomysł (myślę, że w jakichś przyszłych commitach, gdzie będziemy dopracowywać testy, przykłady lub dokumentację, możemy to rozszerzyć).